### PR TITLE
add post

### PR DESCRIPTION
**Pull request by skeydan**

### CODE REVIEW

File: _posts/2021-10-21-segmentation-torch-android/segmentation-torch-android.Rmd [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+    self_contained: false
+    toc: true
+preview: images/segmentation_android.png
+---
+
+```{r setup, include=FALSE}
+knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
+```
+
+In a sense, image segmentation is not that different from image classification. It's just that instead of categorizing an image as a whole, segmentation results in a label for *every single pixel.* And as in image classification, the categories of interest depend on the task: Foreground versus background, say; different types of tissue; different types of vegetation; et cetera.
```
Comment (batpigandme) [inline _posts/2021-10-21-segmentation-torch-android/segmentation-torch-android.Rmd:27]:
```suggestion
In a sense, image segmentation is not that different from image classification. It's just that instead of categorizing an image as a whole, segmentation results in a label for *every single pixel*. And as in image classification, the categories of interest depend on the task: Foreground versus background, say; different types of tissue; different types of vegetation; et cetera.
```

Comment (batpigandme) [reply _posts/2021-10-21-segmentation-torch-android/segmentation-torch-android.Rmd:27]: Semi-obscure rule that the punctuation doesn't get the changed font emphasis unless it's in the middle of said emphasis.

File: _posts/2021-10-21-segmentation-torch-android/segmentation-torch-android.Rmd [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+
+The present post is not the first, on this blog, to treat that topic; and like all prior[^1] ones, it makes use of a U-Net architecture[^2] to achieve its goal. Central characteristics (of this post, not U-Net) are:
```
Comment (batpigandme) [inline _posts/2021-10-21-segmentation-torch-android/segmentation-torch-android.Rmd:29]:
```suggestion
The present post is not the first on this blog to treat that topic; and like all prior[^1] ones, it makes use of a U-Net architecture[^2] to achieve its goal. Central characteristics (of this post, not U-Net) are:
```

Comment (batpigandme) [reply _posts/2021-10-21-segmentation-torch-android/segmentation-torch-android.Rmd:29]: When you say "that topic" do you mean the topic of image segmentation? If no, you lost me 🙂

File: _posts/2021-10-21-segmentation-torch-android/segmentation-torch-android.Rmd [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+
+[^1]: Cf. [Brain image segmentation with torch](https://blogs.rstudio.com/ai/posts/2020-11-30-torch-brain-segmentation/) (using, non-surprisingly, `torch`), and [Image segmentation with U-Net](https://blogs.rstudio.com/ai/posts/2019-08-23-unet/) (using TensorFlow/Keras).
+
+[^2]:  @RonnebergerFB15
+
+1.  It demonstrates how to perform data augmentation for an image segmentation task.
+
+2.  It uses [luz](https://mlverse.github.io/luz/), `torch`'s high-level interface, to train the model.
+
+3.  It [JIT-traces](https://blogs.rstudio.com/ai/posts/2021-08-10-jit-trace-module/) the trained model and saves it for deployment on mobile devices.
```
Comment (batpigandme) [inline _posts/2021-10-21-segmentation-torch-android/segmentation-torch-android.Rmd:39]: I think you need to spell out the words for JIT at least once in this post. Since this is the first mention, it's the logical choice (though I know that's unappealing, given it's in a bullet _and_ a hyphenated term).

File: _posts/2021-10-21-segmentation-torch-android/segmentation-torch-android.Rmd [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+
+4.  It includes proof-of-concept code (though not a discussion) of the saved model being run on Android.
+
+And if you think that this, in itself, is not exciting enough -- our task here is to find cats and dogs. What could be more helpful than a mobile application making sure you can distinguish your cat from the fluffy sofa she's reposing on?
```
Comment (batpigandme) [inline _posts/2021-10-21-segmentation-torch-android/segmentation-torch-android.Rmd:43]:
```suggestion
And if you think that this in itself is not exciting enough -- our task here is to find cats and dogs. What could be more helpful than a mobile application making sure you can distinguish your cat from the fluffy sofa she's reposing on?
```

Comment (batpigandme) [reply _posts/2021-10-21-segmentation-torch-android/segmentation-torch-android.Rmd:43]: 🐈🛋

File: _posts/2021-10-21-segmentation-torch-android/segmentation-torch-android.Rmd [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+library(torchvision)
+library(torchdatasets)
+library(luz)
+
+dir <- "~/.torch-datasets/oxford_pet_dataset"
+
+ds <- oxford_pet_dataset(root = dir)
+```
+
+Images (and corresponding masks) come in different sizes. For training, however, we'll need all of them to be the same size. This can straightforwardly be accomplished by passing in `transform =` and `target_transform =` arguments. But what about data augmentation (basically always a useful measure to take)? Imagine we make use of random flipping. An input image will be flipped -- or not -- according to some probability. But if the image is flipped, the mask better had be, as well! Input and target transformations are not independent, in this case.
```
Comment (batpigandme) [inline _posts/2021-10-21-segmentation-torch-android/segmentation-torch-android.Rmd:70]:
```suggestion
Images (and corresponding masks) come in different sizes. For training, however, we'll need all of them to be the same size. This can be accomplished by passing in `transform =` and `target_transform =` arguments. But what about data augmentation (basically always a useful measure to take)? Imagine we make use of random flipping. An input image will be flipped -- or not -- according to some probability. But if the image is flipped, the mask better had be, as well! Input and target transformations are not independent, in this case.
```

Comment (batpigandme) [reply _posts/2021-10-21-segmentation-torch-android/segmentation-torch-android.Rmd:70]: _straightforwardly_ just sounded like a mouthful, and I think the sentence makes sense without it 👍

File: _posts/2021-10-21-segmentation-torch-android/segmentation-torch-android.Rmd [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+    item <- super$.getitem(i)
+    if (!is.null(self$augmentation)) 
+      self$augmentation(item)
+    else
+      list(x = item$x, y = item$y[1,..])
+  }
+)
+```
+
+All we have to do now is create a custom function that lets us decide on what augmentation to apply once per input-target pair, and then, manually call the respective transformation functions.
```
Comment (batpigandme) [inline _posts/2021-10-21-segmentation-torch-android/segmentation-torch-android.Rmd:120]: Would it make sense to say "what augmentation to apply to each input-target pair"? It's simpler, but it could obscure some nuance of what's being done that I don't fully undersand.

File: _posts/2021-10-21-segmentation-torch-android/segmentation-torch-android.Rmd [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+  raster::plotRGB(img, scale = 1, asp = 1, margins = TRUE)
+  # overlay mask
+  plot(mask, alpha = 0.4, legend = FALSE, axes = FALSE, add = TRUE)
+  
+}
+```
+
+![Learned segmentation masks, overlaid on images from the validation set.](images/pet_segmentation.png)
+
+Now on to running this model "in the wild" (well, sort of).
```
Comment (batpigandme) [inline _posts/2021-10-21-segmentation-torch-android/segmentation-torch-android.Rmd:239]:
```suggestion
Now onto running this model "in the wild" (well, sort of).
```

File: _posts/2021-10-21-segmentation-torch-android/segmentation-torch-android.Rmd [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+traced <- jit_trace(m, x[[1]]$x)
+```
+
+The traced model could now be saved for use with Python or C++, like so:
+
+```{r}
+traced %>% jit_save("traced_model.pt")
+```
+
+However, since we already know we'd like to deploy it on Android, we directly make use of the specialized function `jit_save_for_mobile()` that, additionally, generates bytecode:
```
Comment (batpigandme) [inline _posts/2021-10-21-segmentation-torch-android/segmentation-torch-android.Rmd:262]:
```suggestion
However, since we already know we'd like to deploy it on Android, we make direct use of the specialized function `jit_save_for_mobile()` that, additionally, generates bytecode:
```

File: _posts/2021-10-21-segmentation-torch-android/segmentation-torch-android.Rmd (hunk 3 above)
Comment (skeydan) [reply _posts/2021-10-21-segmentation-torch-android/segmentation-torch-android.Rmd:39]: absolutely, will do!

File: _posts/2021-10-21-segmentation-torch-android/segmentation-torch-android.Rmd (hunk 8 above)
Comment (skeydan) [reply _posts/2021-10-21-segmentation-torch-android/segmentation-torch-android.Rmd:262]: yeah, that doesn't really convey what I had in mind ... a good sign that what I originally wrote was a bad choice! Will find something else

**Comment by skeydan** [comment]
thanks so much Mara!!

File: _posts/2021-10-21-segmentation-torch-android/segmentation-torch-android.Rmd [hunk 9]
```diff
@@ -0,0 +1,283 @@
...
+  initialize = function(..., size, normalize = TRUE, augmentation = NULL) {
+    
+    self$augmentation <- augmentation
+    
+    input_transform <- function(x) {
+      x <- x %>%
+        transform_to_tensor() %>%
+        transform_resize(size) 
+      if (isTRUE(normalize)) x <- x %>%
+        transform_normalize(mean = c(0.485, 0.456, 0.406),
```
Comment (dfalbel) [inline _posts/2021-10-21-segmentation-torch-android/segmentation-torch-android.Rmd:88]: Maybe add a quick comment on why we use these numbers on normalize?

File: _posts/2021-10-21-segmentation-torch-android/segmentation-torch-android.Rmd [hunk 10]
```diff
@@ -0,0 +1,283 @@
...
+train_dl <- dataloader(train_ds, batch_size = 32, shuffle = TRUE)
+valid_dl <- dataloader(valid_ds, batch_size = 32)
+```
+
+### Model training and (visual) evaluation
+
+With `luz`, model training is a matter of two verbs, `setup()` and `fit()`. The learning rate has been determined, for this specific case, using `luz::lr_finder()`; you will likely have to change it when experimenting with different forms of data augmentation (and different data sets).
+
+```{r}
+model <- model %>%
```
Comment (dfalbel) [inline _posts/2021-10-21-segmentation-torch-android/segmentation-torch-android.Rmd:162]: I think it would be useful to have a reference on to where one can find the model definition.

File: _posts/2021-10-21-segmentation-torch-android/segmentation-torch-android.Rmd (hunk 9 above)
Comment (skeydan) [reply _posts/2021-10-21-segmentation-torch-android/segmentation-torch-android.Rmd:88]: will do!

File: _posts/2021-10-21-segmentation-torch-android/segmentation-torch-android.Rmd (hunk 10 above)
Comment (skeydan) [reply _posts/2021-10-21-segmentation-torch-android/segmentation-torch-android.Rmd:162]: Oops??? I have no idea what happened here ... that model definition part went missing! Normally I keep intermediate backups (not on GH), - will add it back tomorrow ...
What luck you spotted this!!!

**Comment by dfalbel** [comment]
Looks great!
